proposals/generate_validating_admission_policies.md

```diff
@@ -0,0 +1,474 @@
+- In case Kyverno policy is deleted, the corresponding VAP and its binding are deleted.
+
+Once validating admission policy and its binding are generated, the API server will handle the resource validation. 
+
+An example of Kyverno policy from which validating admission policy and its binding will be generated:
+```yaml=
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-deployment-replicas
+spec:
+  validationFailureAction: Enforce
+  background: false
+  rules:
+    - name: deployment-replicas
+      match:
+        any:
+        - resources:
+            kinds:
+              - Deployment
+      validate:
+        cel:
+          expressions:
+            - expression: "object.spec.replicas < 5"
+              message:  "Deployment spec.replicas must be less than 5”
+```
+
+## Motivation
+Currently, Kyverno engine supports CEL expressions for resource validation. However, validating admission policy is an alternative solution to validating admission webhooks. Therefore, it's better to make Kyverno generate validating admission policies so that resource validation will be handled by the API server instead.
+
+Users can either let Kyverno engine to validate resources using CEL expressions or let the API server to validate resources and this is done by allowing Kyverno to generate validating admission policies and their bindings from a Kyverno policy containing a CEL subrule.
+
+## Proposal
+- A new controller `validatingadmissionpolicy-controller` will be created. It will watch the following:
+  1. Kyverno policies
+  2. Validating admission policies
+  3. Validating admission policy bindings
+
+- When a policy event happens, the policy key is enqueued. The controller will create/update/delete validating admission policies and their bindings accordingly.
+
+- When a validating admission policy or validating admission policy binding event happens, the policy key is deduced from owner references and it's enqueued. The controller will reset the changes done in either VAP or VAP binding.
+
+- A flag/toggle will be introduced in the controller so that user can either let Kyverno engine handles CEL expressions or let api server handles it by generating VAPs and their bindings.
+
+### Kyverno engine VS API server
+Since CEL expressions can be handled by either Kyverno engine or the API server, Here are the cases where Kyverno engine will handle the resource validations itself:
+1. The flag is disabled in the `validatingadmissionpolicy-controller`. This is done by setting `--generateValidatingAdmissionPolicy` to false.
+2. The flag is enabled and validating admission policies were generated but issues are encountered in processing these policies. This is done by getting the metrics exposed by the API server.
+
+### Limitations
+Validating admission policy and its bindings will be generated for a Kyverno policy that has one `validate.CEL` subrule.
```

Comment:
one or at least one?

Reply:
one `validate.CEL` subrule as we will focus on generating VAPs for a policy that has one rule for now.